network/tcp.go

```diff
@@ -0,0 +1,422 @@
+	if err != nil {
+		return nil, err
+	}
+	return NewRouter(sid, h), nil
+}
+
+// TCPConn is the underlying implementation of
+// Conn using plain Tcp.
+type TCPConn struct {
+	// The name of the endpoint we are connected to.
+	endpoint Address
+
+	// The connection used
+	conn net.Conn
+
+	// closed indicator
+	closed    bool
+	closedMut sync.Mutex
+	// So we only handle one receiving packet at a time
+	receiveMutex sync.Mutex
+	// So we only handle one sending packet at a time
+	sendMutex sync.Mutex
+
+	counterSafe
+}
+
+// NewTCPConn will open a TCPConn to the given address.
+func NewTCPConn(addr Address) (*TCPConn, error) {
+	var err error
+	var conn net.Conn
+	netAddr := addr.NetworkAddress()
+	for i := 0; i < MaxRetryConnect; i++ {
+		conn, err = net.Dial("tcp", netAddr)
+		if err != nil {
+			time.Sleep(WaitRetry)
+		} else {
+			break
+		}
+		time.Sleep(WaitRetry)
+	}
+	if conn == nil {
+		return nil, fmt.Errorf("Could not connect to %s: %s", addr, err)
+	}
+	c := TCPConn{
+		endpoint: addr,
+		conn:     conn,
+	}
+	return &c, err
+}
+
+// Receive calls the receive routine to get the bytes from the connection then
+// it tries to decode the buffer. Returns the Packet with the Msg field decoded
+// or EmptyApplicationPacket and an error if something wrong occured.
+func (c *TCPConn) Receive(ctx context.Context) (nm Packet, e error) {
+	defer func() {
+		if err := recover(); err != nil {
+			e = fmt.Errorf("Error Received message: %v", err)
+			nm = EmptyApplicationPacket
+		}
+	}()
+
+	var am Packet
+	buff, err := c.receive()
+	if err != nil {
+		return EmptyApplicationPacket, err
+	}
+
+	err = am.UnmarshalBinary(buff)
+	if err != nil {
+		return EmptyApplicationPacket, fmt.Errorf("Error unmarshaling message type %s: %s", am.MsgType.String(), err.Error())
+	}
+	am.From = c.Remote()
+	return am, nil
+}
+
+// receive is responsible for getting first the size of the message, then the
+// whole message. It returns the raw message as slice of bytes.
+func (c *TCPConn) receive() ([]byte, error) {
+	c.receiveMutex.Lock()
+	defer c.receiveMutex.Unlock()
+	// First read the size
+	var total Size
+	if err := binary.Read(c.conn, globalOrder, &total); err != nil {
+		return nil, handleError(err)
+	}
+	b := make([]byte, total)
+	var read Size
+	var buffer bytes.Buffer
+	for read < total {
+		// read the size of the next packet
+		n, err := c.conn.Read(b)
+		// if error then quit
+		if err != nil {
+
+			return nil, handleError(err)
+		}
+		// put it in the longterm buffer
+		if _, err := buffer.Write(b[:n]); err != nil {
+			log.Error("Couldn't write to buffer:", err)
+		}
+		read += Size(n)
+		b = b[n:]
+	}
+
+	// set the size read
+	c.updateRx(uint64(read))
+	return buffer.Bytes(), nil
+}
+
+// how many bytes do we write at once on the socket
+// 1400 seems a safe choice regarding the size of a ethernet packet.
+// https://stackoverflow.com/questions/2613734/maximum-packet-size-for-a-tcp-connection
+const maxChunkSize Size = 1400
```

Comment:
I've been mostly copying up things over for the TCP part so I did not really dig into the code but I can't remember why we were doing that in the first place ... and you're right, it seems probably a better choice to leave it out to the TCP stack...